### PR TITLE
Add support for sRGB capable framebuffers.

### PR DESCRIPTION
**Pull request by binary1248**
Implements #175.

### CODE REVIEW

**Comment by LaurentGomila** [comment]
Looks good, except that "RGB" in identifiers should be "Rgb", according to the coding style.

**Comment by binary1248** [comment]
Well... strictly speaking, the whole abbreviation is not RGB but sRGB, so in that sense, the identifiers would have to all be `srgb`.

**Comment by LaurentGomila** [comment]
If we enforce strict PascalCase identifiers, it would be Srgb.

**Comment by binary1248** [comment]
But in camelCase (which we use), it's `srgb`? :stuck_out_tongue:

**Comment by LaurentGomila** [comment]
Yes sorry... we're talking about variable, not class names.

I'd say that ```sRgb``` should be ok then.

**Comment by binary1248** [comment]
Fixed.

**Comment by LaurentGomila** [comment]
:+1:

**Comment by eXpl0it3r** [comment]
How can this best be tested?

**Comment by binary1248** [comment]
Request an sRGB framebuffer and draw some sRGB stuff. You will notice that things are a "bit" different if you just go ahead and request one in the OpenGL example for example. The rendering would naturally have to be adjusted for sRGB if the framebuffer expects it.

**Comment by eXpl0it3r** [comment]
Going to ping a few people who showed interest for this feature in the past and who might be willing to test it: @NicolBolas, @retep998, @bananu7, @elliotpotts, @dezzik

File: src/SFML/Window/Unix/GlxContext.cpp [hunk 1]
```diff
@@ -388,9 +397,19 @@ void GlxContext::updateSettingsFromVisualInfo(XVisualInfo* visualInfo)
...
+    }
+    else
+    {
+        sRgb = 0;
+    }
+
     m_settings.depthBits         = static_cast<unsigned int>(depth);
     m_settings.stencilBits       = static_cast<unsigned int>(stencil);
     m_settings.antialiasingLevel = multiSampling ? samples : 0;
+    m_settings.sRgbCapable       = (sRgb == True);
```
Comment (eXpl0it3r) [inline src/SFML/Window/Unix/GlxContext.cpp]: Where does that `True` come from?

**Comment by eXpl0it3r** [comment]
Are there no other "formats" we might support in the future? I mean, wouldn't it make sense to use an enum instead of a boolean?

File: src/SFML/Window/Unix/GlxContext.cpp (hunk 1 above)
Comment (binary1248) [inline src/SFML/Window/Unix/GlxContext.cpp]: From Xlib via GLX.

**Comment by binary1248** [comment]
sRGB is the only other colour space that I know of which is in widespread use besides "normal RGB". Also, enabling support for sRGB should really be a flag. The specification kind of makes clear that sRGB support is an additional capability of the framebuffer and therefore doesn't replace or gets replaced by anything else.

**Comment by Bromeon** [comment]
> Going to ping a few people who showed interest for this feature in the past and who might be willing to test it: @NicolBolas, @retep998, @bananu7, @elliotpotts, @dezzik

Another attempt :)

**Comment by retep998** [comment]
sRGB is basically the format that everything uses. All images are assumed to be in sRGB (unless specified otherwise). Monitors are calibrated for sRGB. The web dictates sRGB. All this is doing is converting between gamma encoded sRGB and linear RGB so that blending can be performed correctly instead of with a poor approximation. This is __thre__ format to support, therefore in my mind using a boolean is the right thing to do.

This doesn't seem like it handles loading textures as sRGB though. That's an essential counterpart to treating the output framebuffer as sRGB, otherwise when you load in an image and display it to the screen, it'll look wrong.

**Comment by binary1248** [comment]
Added additional support for conversion of texture data from sRGB back to linear so that it can be processed in linear and output back to the framebuffer in sRGB. Also updated the OpenGL example to "demonstrate" its usage (there is no visible difference, as to be expected when not doing e.g. complex lighting).

**Comment by eXpl0it3r** [comment]
Can this be merged?

**Comment by binary1248** [comment]
The only "real" way to test this would be to write e.g. an HDR shader that manipulates the linearized sRGB data of some texture. The whole purpose for all of this is because such algorithms would produce wrong results if the data was not linear.

**Comment by retep998** [comment]
Several tests I can think of:

1. Take a solid gray texture that is srgb(0.5, 0.5, 0.5) and draw it twice using additive blending. If the result is white you're not using sRGB. If the result is merely a light gray srgb(0.68, 0.68, 0.68) then sRGB is working correctly.
2. Take a black and white checkerboard pattern (where the checkers are 1 pixel big) and draw it at 50% scale using linear blending and/or mipmapping so it all blurs together. If the resulting gray is srgb(0.5, 0.5, 0.5) you're not using sRGB. If the result is srgb(0.73, 0.73, 0.73) then sRGB is working correctly.

**Comment by binary1248** [comment]
Test with this then:
```cpp
#include <SFML/Graphics.hpp>
#include <vector>

int main()
{
    // Enable/Disable sRGB here
    bool sRgb = true;

    sf::ContextSettings contextSettings;
    contextSettings.sRgbCapable = sRgb;

    sf::RenderWindow window(sf::VideoMode(800, 600), "sRGB Test", sf::Style::Default, contextSettings);

    sf::Image image;
    image.create(400, 300, sf::Color(127, 127, 127));

    sf::Texture texture1;
    texture1.setSrgb(sRgb);
    texture1.loadFromImage(image);

    sf::Sprite sprite1;
    sprite1.setTexture(texture1);

    std::vector<sf::Color> data(799 * 600, sf::Color::White);
    for(size_t i = 0; i < 799 * 600; i += 2)
        data[i] = sf::Color::Black;

    sf::Texture texture2;
    texture2.setSrgb(sRgb);
    texture2.setSmooth(true);
    texture2.create(799, 600);
    texture2.update(reinterpret_cast<const sf::Uint8*>(&data[0]));

    sf::Sprite sprite2;
    sprite2.setPosition(401, 300);
    sprite2.setTexture(texture2);
    sprite2.setScale(0.5f, 0.5f);

    while(window.isOpen())
    {
        sf::Event event;
        while(window.pollEvent(event))
        {
            if(event.type == sf::Event::Closed)
                return 0;
        }

        window.clear();
        window.draw(sprite1, sf::BlendAdd);
        window.draw(sprite1, sf::BlendAdd);
        window.draw(sprite2);
        window.display();
    }
}
```
It produces the results you would expect with/without sRGB support.

**Comment by eXpl0it3r** [comment]
`sRgb = true`
SRGB

`sRgb = false`
not SRGB

**Comment by mantognini** [comment]
Same here on OS X.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

**Pull request merged**